source/megengine.js

```diff
@@ -41,279 +89,359 @@ megengine.Model = class {
+
+                    if (opr.tensors.length !== 0) {
+                        const initializer = opr.type === 'Host2DeviceCopy' ? undefined : new megengine.Tensor('', opr.tensors[0]);
+                        args.push(new megengine.Argument(name, opr.tensors[0], initializer));
                     }
                     else {
-                        attr_name = meta.varargs + (arg_idx - meta.attributes.length);
-                    }
-                    const rst = process_args(formatTreeDef(arg), inp_idx);
-                    state[attr_name] = rst[0];
-                    inp_idx = rst[1];
-                    arg_idx += 1;
-                }
-                for (let i = 0; i < kwargs.children_defs.length; i++) {
-                    const rst = process_args(formatTreeDef(kwargs.children_defs[i]), inp_idx);
-                    inp_idx = rst[1];
-                    state[kwargs.aux_data[i]] = rst[0];
-                }
-                return state;
-            };
-            const getname = (context, name) => {
-                let rst = name;
-                while (context.get(rst) !== undefined) {
-                    if (rst === context.get(rst)) {
-                        return rst;
-                    }
-                    rst = context.get(rst);
-                }
-                return rst;
-            };
-            const getfullname = (prefix, name) => {
-                return prefix === '' ? name : prefix + '_' + name;
-            };
-            for (const expr of expressions) {
-                const type = expr.__class__.__name__;
-                for (const i of expr.inputs) {
-                    i._fullname = getname(context, getfullname(name_prefix, i._name));
-                }
-                for (const o of expr.outputs) {
-                    o._fullname = getname(context, getfullname(name_prefix, o._name));
-                }
-                switch (type) {
-                    case 'Input': {
-                        break;
+                        args.push(new megengine.Argument(name));
                     }
-                    case 'GetAttr': {
-                        if (expr.outputs[0].__class__.__name__ === 'TensorNode') {
-                            const tensor = parse_getattr(tmodule, expr);
-                            expr.outputs[0].initializer = new megengine.Tensor(expr.name, tensor);
+                    return {name: name, type: type, args: args};
+                };
+
+                const gather_opr_tensor = new Map();
+                let key_id = 0;
+                for (const opr of obj.oprs) {
+                    if (opr.type === 'MultipleDeviceTensorWithFormatHolder') {
+                        for (var id = 0; id < opr.outputs.length; id++) {
+                            const name = obj.middle_tensors[opr.outputs[id]].name;
+                            const tensors = [opr.tensors[id]];
+                            const type = 'ImmutableTensor';
```

Comment:
Most inputs are named `Host2DeviceCopy`, `Convolution` or `ImmutableTensor`. Formats use descriptive names like `weights`, `kernel` or `bias`. If this is not included in the file it should be added via metadata.

Reply:
done